controllers/servingruntime_controller.go

```diff
@@ -509,25 +509,55 @@ func (r *ServingRuntimeReconciler) SetupWithManager(mgr ctrl.Manager,
 }
 
 func (r *ServingRuntimeReconciler) requestsForRuntimes(namespace string,
-	filter func(*kserveapi.ServingRuntime) bool) []reconcile.Request {
+	filter func(string) bool) []reconcile.Request {
 	var opts []client.ListOption
 	if namespace != "" {
 		opts = []client.ListOption{client.InNamespace(namespace)}
 	}
-	list := &kserveapi.ServingRuntimeList{}
-	if err := r.Client.List(context.TODO(), list, opts...); err != nil {
+	runtimes := &kserveapi.ServingRuntimeList{}
+	if err := r.Client.List(context.TODO(), runtimes, opts...); err != nil {
 		r.Log.Error(err, "Error listing ServingRuntimes to reconcile", "namespace", namespace)
 		return []reconcile.Request{}
 	}
-	requests := make([]reconcile.Request, 0, len(list.Items))
-	for i := range list.Items {
-		rt := &list.Items[i]
-		if filter == nil || filter(rt) {
-			requests = append(requests, reconcile.Request{
-				NamespacedName: types.NamespacedName{Name: rt.Name, Namespace: rt.Namespace},
-			})
+
+	srnns := make(map[types.NamespacedName]struct{})
+
+	if r.EnableCSRWatch {
+		csrs := &kserveapi.ClusterServingRuntimeList{}
+		if err := r.Client.List(context.TODO(), csrs); err != nil {
+			return []reconcile.Request{}
+		}
+		for i := range csrs.Items {
+			csr := &csrs.Items[i]
+			if csr.Spec.IsMultiModelRuntime() {
+				list := &corev1.NamespaceList{}
+				if err := r.Client.List(context.TODO(), list); err != nil {
```

Comment:
We should just retrieve the namespaces once before the loop, and only if the `namespace` parameter is empty. Something like:
```go
var namespaces []string
if namespace != "" {
    namespaces = []string{namespace}
} else {
   list := &corev1.NamespaceList{}
   if err := r.Client.List(context.TODO(), list); err != nil {
       // ...
   }
   for j := range list.Items {
        ns := &list.Items[j]
        if filter == nil || filter(ns) {
            namespaces = append(namespaces, ns)
        }
    }
}
```

then just use `namespaces` with the CSR loop.

Reply:
Also no need to call `modelMeshEnabled2`, it's done by the filter.